counsel-ag-popup.el

```diff
@@ -0,0 +1,137 @@
+
+;; Copyright (C) 2019  Eder Elorriaga
+
+;; Author: Eder Elorriaga <gexplorer8@gmail.com>
+;; Keywords: convenience, matching, tools
+
+;; This program is free software; you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation, either version 3 of the License, or
+;; (at your option) any later version.
+
+;; This program is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+;;; Commentary:
+
+;; Interactive search with ag.
+;; To enable, put this code in your init file:
+;; (use-package counsel-ag-popup
+;;  :after counsel
+;;  :bind ("C-c s" . counsel-ag-popup))
+
+;;; Code:
+
+(require 'counsel)
+(require 'transient)
+
+(eval-when-compile
+  (require 'subr-x))
+
+(defun counsel-ag-popup-search-here ()
+  "Ag search in current directory."
+  (interactive)
+  (counsel-ag-popup-search default-directory))
+
+(defun counsel-ag-popup-search (directory)
+  "Ag search in DIRECTORY."
+  (interactive "DDirectory: ")
+  (let ((ag-args (counsel-ag-popup--map-args (transient-args 'counsel-ag-popup))))
+    (counsel-ag "" directory ag-args)))
+
+(defun counsel-ag-popup--map-args (transient-args)
+  "Convert TRANSIENT-ARGS to a string of args."
+  (mapconcat
+   (lambda (arg)
+     (if (listp arg)
+         (let ((args (last arg (- (length arg) 1))))
+           (mapconcat (lambda (x) (concat "--" x)) args " "))
+       arg))
+   transient-args
+   " "))
+
+(defun counsel-ag-popup-read-file-types (prompt initial-input history)
+  "Prompt for Ag file type with PROMPT INITIAL-INPUT HISTORY."
+  (completing-read-multiple
+   prompt
+   '("actionscript" "ada" "asciidoc" "asm" "batch" "bitbake" "bro" "cc" "cfmx"
+     "chpl" "clojure" "coffee" "cpp" "crystal" "csharp" "css" "cython" "delphi"
+     "dot" "ebuild" "elisp" "elixir" "elm" "erlang" "factor" "fortran" "fsharp"
+     "gettext" "glsl" "go" "groovy" "haml" "handlebars" "haskell" "haxe" "hh"
+     "html" "ini" "ipython" "jade" "java" "js" "json" "jsp" "julia" "kotlin"
+     "less" "liquid" "lisp" "log" "lua" "m4" "make" "mako" "markdown" "mason"
+     "matlab" "mathematica" "md" "mercury" "nim" "nix" "objc" "objcpp" "ocaml"
+     "octave" "org" "parrot" "perl" "php" "pike" "plist" "plone" "proto" "puppet"
+     "python" "qml" "racket" "rake" "restructuredtext" "rs" "r" "rdoc" "ruby"
+     "rust" "salt" "sass" "scala" "scheme" "shell" "smalltalk" "sml" "sql"
+     "stylus" "swift" "tcl" "tex" "tt" "toml" "ts" "twig" "vala" "vb" "velocity"
+     "verilog" "vhdl" "vim" "wix" "wsdl" "wadl" "xml" "yaml")
+   nil nil
+   initial-input
```

Comment:
Why provide for the `initial-input` argument, which is generally discouraged in completion for being intrusive, and not the `def` argument?

Reply:
I didn't know that `initial-input` was discouraged. I just checked the doc for `completing-read` and read about the `initial-input` vs `def` thing.o

The reason why it's like this is that I got it from `magit-diff.el`
```elisp
(defun magit-read-files (prompt initial-input history)
  (magit-completing-read-multiple* prompt
                                   (magit-list-files)
                                   nil nil
                                   (or initial-input (magit-file-at-point))
                                   history))
```

I'll change `initial-input` for `def` and see if it works.